### ISSUE TITLE
REOPEN: ntopng - "database schema is being updated..."

### ISSUE THREAD

**Issue by njcwotx** [ntop/ntopng#741]
I opened a case last week but it was closed before I think the issue is understood.

Another restart of the ntopng.service today and another very long "MySQL schema update. Altering table flowsv4: changing engine from InnoDB to MyISAM." in the ntopng.log.

I did no software update, but every time I restart the service the historical database goes though a schema update. The larger the database grows, the longer this takes. It was at 52GB before the restart and took 58 minutes to do the schema update. This was a 4-5 day old database, I can only imagine the size it will be after a month. I set the history to 3 days for now, but I imagine I will have to set longer very soon.

As you mentioned, if there is no schema update in the version I am using, why does it have to update the schema after every restart? I can understand one schema update after an upgrade, but each time I restart a service seems to me uneccesary.

original issue:
https://github.com/ntop/ntopng/issues/732

**Comment by simonemainardi** [comment]
according to your report, ntopng says "flowsv4: changing engine from InnoDB to MyISAM" after every restart. Since ntopng checks for the table engine _before_ doing the change, it is likely that ntopng detects the InnoDB engine, even if it should be MyISAM.

Please, log into the database and do the following query:

```
mysql> show table status;
+---------+--------+---------+------------+-------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+------------+-----------------+----------+----------------+---------+
| Name    | Engine | Version | Row_format | Rows  | Avg_row_length | Data_length | Max_data_length | Index_length | Data_free | Auto_increment | Create_time         | Update_time         | Check_time | Collation       | Checksum | Create_options | Comment |
+---------+--------+---------+------------+-------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+------------+-----------------+----------+----------------+---------+
| flowsv4 | MyISAM |      10 | Dynamic    | 86854 |            356 |    30954768 |               0 |      6966272 |         0 |          87374 | 2016-09-06 17:08:40 | 2016-09-13 16:10:52 | NULL       | utf8_general_ci |     NULL | partitioned    |         |
| flowsv6 | MyISAM |      10 | Dynamic    | 39499 |            261 |    10339912 |               0 |      2385920 |         0 |          39506 | 2016-09-06 17:08:40 | 2016-09-13 16:10:22 | NULL       | utf8_general_ci |     NULL | partitioned    |         |
+---------+--------+---------+------------+-------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+------------+-----------------+----------+----------------+---------+
2 rows in set (0.00 sec)

mysql>


```

double check tables engine both after ntopng has completed the migration and right before starting ntopng. We should make sure that tables are correctly migrated to MyISAM.

**Comment by njcwotx** [comment]
# showtable after starting ntopng and the schema update completed.

MariaDB [(none)]> use ntopng;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [ntopng]> show table status;
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+
| Name    | Engine | Version | Row_format | Rows      | Avg_row_length | Data_length | Max_data_length | Index_length | Data_free | Auto_increment | Create_time         | Update_time         | Check_time          | Collation       | Checksum | Create_options | Comment |
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+
| flowsv4 | MyISAM |      10 | Dynamic    | 121636150 |            337 | 40996441376 |               0 |  13045075968 |         0 |      152414642 | 2016-09-13 13:47:48 | 2016-09-13 16:27:39 | 2016-09-13 14:44:49 | utf8_general_ci |     NULL | partitioned    |         |
| flowsv6 | MyISAM |      10 | Dynamic    |    320937 |            284 |    91236992 |               0 |     43786240 |         0 |         348818 | 2016-09-13 14:44:59 | 2016-09-13 16:27:36 | 2016-09-13 14:45:06 | utf8_general_ci |     NULL | partitioned    |         |
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+
2 rows in set (0.02 sec)

# MariaDB [ntopng]>

## after stopping ntopng and doing a show table

MariaDB [(none)]> use ntopng;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [ntopng]> show table status;
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+
| Name    | Engine | Version | Row_format | Rows      | Avg_row_length | Data_length | Max_data_length | Index_length | Data_free | Auto_increment | Create_time         | Update_time         | Check_time          | Collation       | Checksum | Create_options | Comment |
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+
| flowsv4 | MyISAM |      10 | Dynamic    | 121728004 |            337 | 41027569524 |               0 |  13052838912 |         0 |      152506496 | 2016-09-13 13:47:48 | 2016-09-13 16:30:42 | 2016-09-13 14:44:49 | utf8_general_ci |     NULL | partitioned    |         |
| flowsv6 | MyISAM |      10 | Dynamic    |    321085 |            284 |    91279000 |               0 |     43802624 |         0 |         348966 | 2016-09-13 14:44:59 | 2016-09-13 16:30:42 | 2016-09-13 14:45:06 | utf8_general_ci |     NULL | partitioned    |         |
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+

## 2 rows in set (0.00 sec)

# ntopng service started...awaiting the schema update to complete...

MariaDB [(none)]> use ntopng;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [ntopng]> show table status;
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+
| Name    | Engine | Version | Row_format | Rows      | Avg_row_length | Data_length | Max_data_length | Index_length | Data_free | Auto_increment | Create_time         | Update_time         | Check_time          | Collation       | Checksum | Create_options | Comment |
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+
| flowsv4 | MyISAM |      10 | Dynamic    | 121728004 |            337 | 41027569524 |               0 |  13052838912 |         0 |      152506496 | 2016-09-13 13:47:48 | 2016-09-13 16:30:42 | 2016-09-13 14:44:49 | utf8_general_ci |     NULL | partitioned    |         |
| flowsv6 | MyISAM |      10 | Dynamic    |    321085 |            284 |    91279000 |               0 |     43802624 |         0 |         348966 | 2016-09-13 14:44:59 | 2016-09-13 16:30:42 | 2016-09-13 14:45:06 | utf8_general_ci |     NULL | partitioned    |         |
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+

# 2 rows in set (0.01 sec)

will post in an hour or so after the update runs.

**Comment by njcwotx** [comment]
here is the status right after the schema update was reapplied..

MariaDB [(none)]> use ntopng;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [ntopng]> show table status;
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+
| Name    | Engine | Version | Row_format | Rows      | Avg_row_length | Data_length | Max_data_length | Index_length | Data_free | Auto_increment | Create_time         | Update_time         | Check_time          | Collation       | Checksum | Create_options | Comment |
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+
| flowsv4 | MyISAM |      10 | Dynamic    | 122673707 |            337 | 41347469332 |               0 |  13077167104 |         0 |      153452199 | 2016-09-13 16:31:48 | 2016-09-13 18:04:15 | 2016-09-13 17:28:09 | utf8_general_ci |     NULL | partitioned    |         |
| flowsv6 | MyISAM |      10 | Dynamic    |    323412 |            284 |    91940708 |               0 |     43972608 |         0 |         351293 | 2016-09-13 17:28:17 | 2016-09-13 18:04:12 | 2016-09-13 17:28:22 | utf8_general_ci |     NULL | partitioned    |         |
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-----------------+----------+----------------+---------+
2 rows in set (0.01 sec)

MariaDB [ntopng]>

**Comment by simonemainardi** [comment]
So the table engine is always MyISAM and this is OK. Maybe the point is that you are using mariadb and not MySQL.

I have tested that with mariadb on centos7 and I don't get updates every time I restart the service. Please report the exact OS and ntopng version used.

Also try to do the following query:

```
MariaDB [(none)]> select 1 from information_schema.TABLES where TABLE_SCHEMA = 'ntopng'  and TABLE_NAME='flowsv4' and ENGINE='InnoDB';
```

ntopng uses the query above to see if the engine is InnoDB. Thus, if the return is not empty, ntopng performs the update. So I would expect and `Empty Set` returned but it is worth checking that as well.

**Comment by colinvv2** [comment]
Mine does this after every restart... and it looks like MySQL is set as expected for MyISAM?

`mysql> select 1 from information_schema.TABLES where TABLE_SCHEMA = 'ntopng'  and TABLE_NAME='flowsv4' and ENGINE='InnoDB';
Empty set (0.00 sec)

mysql> use ntopng;
Reading table information for completion of table and column names

mysql> show table status;
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-------------------+----------+----------------+---------+
| Name    | Engine | Version | Row_format | Rows      | Avg_row_length | Data_length | Max_data_length | Index_length | Data_free | Auto_increment | Create_time         | Update_time         | Check_time          | Collation         | Checksum | Create_options | Comment |
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-------------------+----------+----------------+---------+
| flows   | MyISAM |      10 | Dynamic    |  15442662 |             51 |   802782788 | 281474976710655 |   1244312576 |         0 |       15442663 | 2016-08-27 07:47:24 | 2016-09-14 00:17:31 | 2016-08-27 07:47:24 | latin1_swedish_ci |     NULL |                |         |
| flowsv4 | MyISAM |      10 | Dynamic    | 121099240 |            337 | 40905973620 |               0 |   7351832576 | 563513380 |      122904629 | 2016-09-05 10:41:06 | 2016-09-13 23:39:24 | 2016-09-05 12:02:59 | utf8_general_ci   |     NULL | partitioned    |         |
| flowsv6 | MyISAM |      10 | Dynamic    |         0 |              0 |           0 |               0 |       131072 |         0 |              7 | 2016-09-05 12:03:00 | 2016-09-05 12:03:00 | NULL                | utf8_general_ci   |     NULL | partitioned    |         |
+---------+--------+---------+------------+-----------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+---------------------+---------------------+-------------------+----------+----------------+---------+
3 rows in set (0.00 sec)`

**Comment by simonemainardi** [comment]
please, try to update to the latest stable that I've added an extra check in the mysql query. If that doesn't solve the issue, can you send the credentials to do a remote session? you can send them privately at mainardi ntop org

**Comment by colinvv2** [comment]
Updated to latest release...

Get:9 http://www.nmon.net/apt-stable/14.04/ x64/ ntopng 2.4.160913-1653 [4,174 kB]
Get:10 http://www.nmon.net/apt-stable/14.04/ x64/ nprobe 7.4.160913-5333 [1,135 kB]
Get:11 http://www.nmon.net/apt-stable/14.04/ x64/ n2disk 2.6.160913-4666 [957 kB]
Get:12 http://www.nmon.net/apt-stable/14.04/ x64/ pfring 6.4.1-798 [1,476 kB]
Get:13 http://www.nmon.net/apt-stable/14.04/ all/ nbox 2.4-511 [1,661 kB]
Get:14 http://www.nmon.net/apt-stable/14.04/ all/ ntopng-data 2.4.160913 [21.4 MB]

Now waiting for MySQL 'ALTER TABLE `flowsv4` ENGINE='MyISAM' to complete... then I will restart server again.

ID      Username    Client host     Database    Mode    Connected       Running query
40  ntopng  localhost   ntopng  Query   02:04:03    ALTER TABLE `flowsv4` ENGINE='MyISAM'

**Comment by colinvv2** [comment]
Yeah! No DB alter at start this time... and I am seeing flows come through ntopng.

**Comment by simonemainardi** [comment]
cool, thanks for reporting ;)

**Comment by njcwotx** [comment]
my reposoitory pointed to the /centos-stable folder and updating the ntopng-2.4.160913-1307.x86_64.rpm via yum update did not work and subsequent restart took 90 minutes.  I found the /rpm7 folder with ntopng-2.5.160916-1560.x86_64.rpm package and used this version manually and the db schema update did not reoccur.

Thanks for fixing this up.

**Comment by colinvv2** [comment]
simonemainardi,
Since the last upgrade, my ntopng has stopped running twice on it's own. Last night at exactly 1:00am and the night before at exactly 1:00am. Strange??? I will post log to new ticket.

**Comment by simonemainardi** [comment]
colin this has been fixed. see #751
